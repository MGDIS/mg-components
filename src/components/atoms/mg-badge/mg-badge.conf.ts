/**
 * List of all possibles variants
 */
export const variants = ['info', 'primary', 'secondary', 'success', 'warning', 'danger'] as const;

/**
 * Variant type from variants
 */
export type BadgeVariantType = (typeof variants)[number];
