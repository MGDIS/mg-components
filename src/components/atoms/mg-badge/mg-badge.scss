@mixin badge-variant($name, $backgroundColor, $color) {
  &.mg-badge--#{$name} {
    // Decoration
    background-color: hsl(var(--color-#{$backgroundColor}));
    border-color: hsl(var(--color-#{$backgroundColor}));
    // Font
    color: hsl(var(--color-#{$color}));

    &.mg-badge--outline {
      // Decoration
      background-color: transparent;
      // Font
      color: hsl(var(--color-#{$backgroundColor}));
    }
  }
}

.mg-badge {
  $vertical-padding: 0.3rem;
  $line-height: var(--mg-badge-font-size);
  // Display
  padding: $vertical-padding 0.7em;
  min-height: calc($vertical-padding * 2 + $line-height);
  box-sizing: border-box;
  // Decoration
  border-radius: var(--mg-badge-border-radius);
  border: 0.1rem solid;
  // Font
  font-size: var(--mg-badge-font-size);
  line-height: $line-height;
  font-weight: 600;
  text-align: center;

  // Variant
  @include badge-variant('primary', 'dark', 'light');
  @include badge-variant('secondary', 'light', 'dark');
  @include badge-variant('success', 'success', 'dark');
  @include badge-variant('warning', 'warning', 'dark');
  @include badge-variant('danger', 'danger', 'light');
  @include badge-variant('info', 'info', 'light');
}

.mg-badge,
mg-badge {
  display: inline-block;
  vertical-align: middle;
}
