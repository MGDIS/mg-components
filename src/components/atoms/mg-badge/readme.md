## Design

The badge is always placed on top of, or next to the element it is for.

The badge displays a number (can be followed by the `+` character) or a punctuation character.

## Specs

![](./mg-badge/doc/img/mg-badge-specs.png)

## Theming

![](./mg-badge/doc/img/mg-badge-styles.png)

## CSS Variables

If needed some [variables](./?path=/story/css-variables--page) are available to customize the component:

- `--mg-badge-size`: define badge height and min-width, default: `1.6rem`
- `--mg-badge-font-size`: define badge font size, default: `1.1rem`

<!-- Auto Generated Below -->


## Properties

| Property             | Attribute | Description                                                        | Type               | Default       |
| -------------------- | --------- | ------------------------------------------------------------------ | ------------------ | ------------- |
| `label` _(required)_ | `label`   | Badge label. Include short description. Required for accessibility | `string`           | `undefined`   |
| `outline`            | `outline` | Define if button is using outline style                            | `boolean`          | `undefined`   |
| `value` _(required)_ | `value`   | Badge value                                                        | `number \| string` | `undefined`   |
| `variant`            | `variant` | Define button variant                                              | `string`           | `variants[0]` |


## Dependencies

### Used by

 - [mg-tabs](../../molecules/mg-tabs)

### Graph
```mermaid
graph TD;
  mg-tabs --> mg-badge
  style mg-badge fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
