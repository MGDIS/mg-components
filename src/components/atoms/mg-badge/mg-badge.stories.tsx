import { h } from '@stencil/core';
import { variants } from './mg-badge.conf';

export default {
  component: 'mg-badge',
  title: 'Atoms/mg-badge',
  argTypes: {
    variant: {
      options: variants,
      control: { type: 'select' },
      table: {
        defaultValue: { summary: variants[0] },
      },
    },
    value: {
      control: { type: 'text' },
    },
  },
};

const Template = args => {
  // Extract slot so it won't be render as an attribute
  // return element
  return <mg-badge {...args}></mg-badge>;
};

export const MgBadge = Template.bind({});
MgBadge.args = {
  value: '99',
  label: 'unread messages',
  variant: variants[0],
  outline: false,
};
