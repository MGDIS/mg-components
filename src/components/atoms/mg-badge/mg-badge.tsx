import { Component, h, Prop, State, Watch } from '@stencil/core';
import { variants, BadgeVariantType } from './mg-badge.conf';
import { ClassList } from '../../../utils/components.utils';
@Component({
  tag: 'mg-badge',
  styleUrl: 'mg-badge.scss',
  shadow: true,
})
export class MgBadge {
  /************
   * Internal *
   ************/

  // Classes
  private classOutline = `mg-badge--outline`;

  /**************
   * Decorators *
   **************/

  /**
   * Badge value
   */
  @Prop({ mutable: true }) value!: string | number;
  @Watch('value')
  validateValue(newValue: MgBadge['value']): void {
    if (`${newValue}`.match(/^(\d+\+*|[?*!a-z])$/i) === null) {
      throw new Error('<mg-badge> prop "value" must be interger or ponctuation character.');
    }
  }

  /**
   * Badge label. Include short description.
   * Required for accessibility
   */
  @Prop() label!: string;
  @Watch('label')
  validateLabel(newValue: MgBadge['label']): void {
    if (typeof newValue !== 'string' || newValue.trim() === '') {
      throw new Error('<mg-badge> prop "label" is required.');
    }
  }

  /**
   * Define button variant
   */
  @Prop() variant?: BadgeVariantType = variants[0]; // info
  @Watch('variant')
  validateVariant(newValue: MgBadge['variant'], oldValue?: MgBadge['variant']): void {
    if (!variants.includes(newValue)) {
      throw new Error(`<mg-badge> prop "variant" must be one of : ${variants.join(', ')}.`);
    } else {
      if (oldValue !== undefined) {
        this.classList.delete(`mg-badge--${oldValue}`);
      }
      this.classList.add(`mg-badge--${newValue}`);
    }
  }

  /**
   * Define if button is using outline style
   */
  @Prop() outline: boolean;
  @Watch('outline')
  validateOutline(newValue: MgBadge['outline']): void {
    if (newValue) this.classList.add(this.classOutline);
    else this.classList.delete(this.classOutline);
  }

  /**
   * Component classes
   */
  @State() classList: ClassList = new ClassList(['mg-badge']);

  /**
   * Check if props are well configured on init
   *
   * @returns {void}
   */
  componentWillLoad(): void {
    this.validateValue(this.value);
    this.validateLabel(this.label);
    this.validateVariant(this.variant);
    this.validateOutline(this.outline);
  }

  /**
   * Render
   *
   * @returns {HTMLElement} HTML Element
   */
  render(): HTMLElement {
    return (
      <span class={this.classList.join()}>
        {this.value}
        <span class="sr-only">{this.label}</span>
      </span>
    );
  }
}
