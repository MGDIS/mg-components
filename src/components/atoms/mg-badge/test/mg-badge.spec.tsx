import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { MgBadge } from '../mg-badge';
import { variants } from '../mg-badge.conf';

const getPage = args =>
  newSpecPage({
    components: [MgBadge],
    template: () => <mg-badge {...args}></mg-badge>,
  });

describe('mg-badge', () => {
  describe.each(variants)('variant %s', variant => {
    test.each([true, false])('outiline %s', async outline => {
      const { root } = await getPage({ variant, outline, value: 1, label: 'batman' });
      expect(root).toMatchSnapshot();
    });
  });

  test.each([1, 100, '!', '?'])('value %s', async value => {
    const { root } = await getPage({ value, label: 'batman' });
    expect(root).toMatchSnapshot();
  });

  describe('errors', () => {
    test.each(['', 'batman', undefined])('Should throw error, case invalid variant prop', async variant => {
      try {
        await getPage({ variant, value: 1, label: 'batman' });
      } catch (err) {
        expect(err.message).toContain('<mg-badge> prop "variant" must be one of : ');
      }
    });

    test.each(['', 'batman', '+', '99?'])('Should throw error, case invalid value prop', async value => {
      try {
        await getPage({ value, label: 'batman' });
      } catch (err) {
        expect(err.message).toContain('<mg-badge> prop "value" must be interger or ponctuation character.');
      }
    });
  });
});
