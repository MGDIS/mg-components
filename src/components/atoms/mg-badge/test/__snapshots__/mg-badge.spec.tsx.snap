// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mg-badge variant danger outiline false 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-badge mg-badge--danger" part="badge-content">
      <slot></slot>
    </span>
  </mock:shadow-root>
  danger
</mg-badge>
`;

exports[`mg-badge variant danger outiline true 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-badge mg-badge--danger mg-badge--outline" part="badge-content">
      <slot></slot>
    </span>
  </mock:shadow-root>
  danger
</mg-badge>
`;

exports[`mg-badge variant info outiline false 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-badge mg-badge--info" part="badge-content">
      <slot></slot>
    </span>
  </mock:shadow-root>
  info
</mg-badge>
`;

exports[`mg-badge variant info outiline true 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-badge mg-badge--info mg-badge--outline" part="badge-content">
      <slot></slot>
    </span>
  </mock:shadow-root>
  info
</mg-badge>
`;

exports[`mg-badge variant primary outiline false 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-badge mg-badge--primary" part="badge-content">
      <slot></slot>
    </span>
  </mock:shadow-root>
  primary
</mg-badge>
`;

exports[`mg-badge variant primary outiline true 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-badge mg-badge--outline mg-badge--primary" part="badge-content">
      <slot></slot>
    </span>
  </mock:shadow-root>
  primary
</mg-badge>
`;

exports[`mg-badge variant secondary outiline false 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-badge mg-badge--secondary" part="badge-content">
      <slot></slot>
    </span>
  </mock:shadow-root>
  secondary
</mg-badge>
`;

exports[`mg-badge variant secondary outiline true 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-badge mg-badge--outline mg-badge--secondary" part="badge-content">
      <slot></slot>
    </span>
  </mock:shadow-root>
  secondary
</mg-badge>
`;

exports[`mg-badge variant success outiline false 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-badge mg-badge--success" part="badge-content">
      <slot></slot>
    </span>
  </mock:shadow-root>
  success
</mg-badge>
`;

exports[`mg-badge variant success outiline true 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-badge mg-badge--outline mg-badge--success" part="badge-content">
      <slot></slot>
    </span>
  </mock:shadow-root>
  success
</mg-badge>
`;

exports[`mg-badge variant warning outiline false 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-badge mg-badge--warning" part="badge-content">
      <slot></slot>
    </span>
  </mock:shadow-root>
  warning
</mg-badge>
`;

exports[`mg-badge variant warning outiline true 1`] = `
<mg-badge>
  <mock:shadow-root>
    <span class="mg-badge mg-badge--outline mg-badge--warning" part="badge-content">
      <slot></slot>
    </span>
  </mock:shadow-root>
  warning
</mg-badge>
`;
